rust/azure_iot_operations_services/src/azure_device_registry/client.rs

```diff
@@ -0,0 +1,983 @@
+    /// if there are any underlying errors from the AIO RPC protocol.
+    ///
+    /// [`struct@Error`] of kind [`ObservationError`](ErrorKind::ObservationError)
+    /// if notification response failed.
+    ///
+    /// [`struct@Error`] of kind [`DuplicateObserve`](ErrorKind::DuplicateObserve)
+    /// if duplicate assets are being observed.
+    pub async fn observe_asset_update(
+        &self,
+        aep_name: String,
+        asset_name: String,
+        timeout: Duration,
+    ) -> Result<AssetObservation, Error> {
+        // TODO Right now using aep_name + asset_name as the key for the dispatcher, consider using tuple
+        let receiver_id = aep_name.clone() + "~" + &asset_name;
+
+        let rx = self
+            .asset_update_event_telemetry_dispatcher
+            .register_receiver(receiver_id.clone())
+            .map_err(|_| Error(ErrorKind::DuplicateObserve))?;
+
+        let notification_payload = NotifyOnAssetUpdateRequestPayload {
+            notification_request: NotifyOnAssetUpdateRequestSchema {
+                asset_name: asset_name.clone(),
+                notification_message_type: NotificationMessageType::On,
+            },
+        };
+
+        let command_request = rpc_command::invoker::RequestBuilder::default()
+            .topic_tokens(HashMap::from([("aepName".to_string(), aep_name)]))
+            .payload(notification_payload)
+            .map_err(|e| Error(ErrorKind::SerializationError(e.to_string())))?
+            .timeout(timeout)
+            .build()?;
+
+        let result = self
+            .notify_on_asset_update_command_invoker
+            .invoke(command_request)
+            .await;
+
+        match result {
+            Ok(response) => {
+                if let NotificationResponse::Accepted = response.payload.notification_response {
+                    Ok(AssetObservation {
+                        name: asset_name,
+                        receiver: rx,
+                    })
+                } else {
+                    // TODO Check error kind - another kind needs to be incldued ?
+                    Err(Error(ErrorKind::ObservationError(
+                        ("Notification Response Failed").to_string(),
+                    )))
+                }
+            }
+            Err(e) => {
+                // If the observe request wasn't successful, remove it from our dispatcher
+                if self
+                    .asset_update_event_telemetry_dispatcher
+                    .unregister_receiver(&receiver_id)
+                {
+                    log::debug!(
+                        "AEP and Asset combination removed from observed list: {receiver_id:?}"
+                    );
+                } else {
+                    log::debug!("AEP and Asset combination not in observed list: {receiver_id:?}");
+                }
+                Err(Error(ErrorKind::AIOProtocolError(e)))
```

Comment:
There seems to be an inconsistency in how we're doing errors here - It's reasonable to do this explicit error instantiation if we don't see the ErrorKind as an error itself (my preference), but then some places, functions like `ErrorKind::from` get used, indicating the error kind is itself also an error.

If the `ErrorKind` is itself an error, then we could be doing more automatic conversions, such as here with `AIOProtocolError`. So which is it?